docs/architecture/adr-040-storage-and-smt-state-commitments.md

```diff
@@ -0,0 +1,118 @@
+
++ Each object select requires a tree traversal from the root
++ Each edge traversal requires a DB query (nodes are not stored in a memory)
++ Creating snapshots is [expensive](https://github.com/cosmos/cosmos-sdk/issues/7215#issuecomment-684804950). It takes about 30 seconds to export less than 100 MB of state (as of March 2020).
++ Updates in IAVL may trigger tree reorganization and possible O(log(n)) hashes re-computation, which can become a CPU bottleneck.
++ The leaf structure is pretty expensive: it contains the `(key, value)` pair, additional metadata such as height, version. The entire node is hashed, and that hash is used as the key in the underlying database, [ref](https://github.com/cosmos/iavl/blob/master/docs/node/node.md
+).
+
+
+Moreover, the IAVL project lacks a support and we already see better, well adapted alternatives. Instead of optimizing the IAVL, we are looking in another solutions for both storage and state commitments.
+
+
+## Decision
+
+We propose separate the concerns of state commitment (**SC**), needed for consensus, and state storage (**SS**), needed for state machine. Finally we replace IAVL with LazyLedger SMT.
+
+
+### Decouple state commitment from storage
+
+Separation of storage and SMT will allow a specialization in terms of various optimization patterns.
+
+SMT will use it's own storage (could use the same database underneath) from the state machine store. For every `(key, value)` pair, the SMT will store `hash(key)` in a path and `hash(key, value)` in a leaf.
+
+For data access we propose 2 additional KV buckets:
+1. B1: `key → value`: the principal object storage, used by a state machine, behind the SDK `KVStore` interface.
+2. B2: `hash(key, value) → key`: an index needed to extract a value (through: B2 -> B1) having a only a Merkle Path. Recall that SMT will store `hash(key, value)` in it's leafs.
+3. we could use more buckets to optimize the app usage if needed.
+
+Above, we propose to use KV DB. However, for state machine we could use RDBMS, which we discuss below.
+
+
+### Requirements
+
+State Storage requirements:
++ range queries
++ quick (key, value) access
++ creating a snapshot
+
+State Commitment requirements:
++ fast updates
++ path length should be short
++ creating a snapshot
++ pruning
+
+
+### LazyLedger SMT for State Commitment
+
+A Sparse Merkle tree is based on the idea of a complete Merkle tree of an intractable size. The assumption here is that as the size of the tree is intractable, there would only be a few leaf nodes with valid data blocks relative to the tree size, rendering the tree as sparse.
+
+TODO:
++ describe pruning and version management
+
+### Snapshots
+
+One of the Stargate core features are snapshots and fast sync. Currently this feature is implemented through IAVL.
```

Comment:
I can't see a way to use snapshots/versions with BoltDB API. BoltDB provides "snapshot isolation", but do not support explicit creation/usage of snapshots.

Reply:
That's a good point. We should double-check that the dbs mentioned below actually provide the desired snapshotting mechanism.

Reply:
Thanks for checking. I'm not completely sure. We can check with the hashicorp team. I found this:
+ "If a long running read transaction (for example, a snapshot transaction) is needed, you might want to set DB.InitialMmapSize to a large enough value to avoid potential blocking of write transaction. " [source](https://pkg.go.dev/github.com/boltdb/bolt#DB.Begin)
+ This _sovled_ issue could be relevant, because waypoint is using boltDB and it support snapshots: https://github.com/hashicorp/waypoint/issues/682

Reply:
I'm not sure if we should spend a time here to discuss the backend. I mean - we need to verify the proposals, but we should have a separate discussion about the the "recommended" DB backend. 
So, what do you think about shifting the snapshot functionality to DB?

Reply:
I'm wondering if using database snapshots to access old state roots is overkill, or even necessary for SMT.

By default, you can already access the old state roots in the LazyLedger SMT implementation, because the tree isn't garbage collected currently. Once garbage collection is added, it could be configured to only garbage collect state roots older than a certain version, which would be equivalent to snapshots, no?

Reply:
> "If a long running read transaction (for example, a snapshot transaction) is needed, you might want to set DB.InitialMmapSize to a large enough value to avoid potential blocking of write transaction. " 

I saw this before, and I'm pretty sure it's about [app-level logic](https://github.com/hashicorp/waypoint/blob/main/internal/server/singleprocess/state/snapshot.go#L51) executing app-level transaction (iteration over entire KV-store) to generate app-level snapshots. 

> I'm wondering if using database snapshots to access old state roots is overkill, or even necessary for SMT.

I think we discuss snapshots because of the "storage" part of proposal not "state commitment" part. For SMT, I also don't see a reason to use DB-level snapshots.

> So, what do you think about shifting the snapshot functionality to DB?

I agree that it is more than reasonable to use capabilities of DB to ensure we can access&manage previous versions of state.

Reply:
Good point. 
We need to have old versions of SMT for snapshots. So, I see two options here:
1. Whenever we schedule a snapshot, we do it in a same way as for state snapshots: using DB builtin feature. If we need to create a new node, and orphan some other node, then we immediately remove orphan nodes. So, effectively, we won't have old roots because we will remove them immediately.
2. Extend the current LL SMT and add pruning.

PS: I assume in your current implementation, you don't update nodes - instead you create a new one, right? This allows you to keep all old versions. Why do you do that, instead of updating nodes?

Reply:
We have to add pruning to LL SMT anyway, otherwise the state will keep growing.

>PS: I assume in your current implementation, you don't update nodes - instead you create a new one, right? This allows you to keep all old versions. Why do you do that, instead of updating nodes?

Nodes are stored in a key => value store where the key is a hash of the node, and the value is the preimage of the hash. Assuming the hashes are collision resistant, you can't "update" a node since the tree is immutable - you can only "create" a new tree with a new root.

Reply:
But you have a path to the updated Leaf, so you can use it to remove old nodes:
1. Given `key`, we traverse the tree with `path=hash(key)`
2. While going down to a node, we can remember all nodes on the path
3. When going back and updating (or creating new ones) internal nodes, we can remove old node - we use the path we constructed in the step `2.`